### PR TITLE
Cleanup Windows warning suppression in CMake and fix some warnings in the source code

### PR DESCRIPTION
**Pull request by cyyever**
This PR do two things:
1. It moves some Windows warning suppression from various CMake files into the main CMakeList.txt, following the conventions of gcc and clang.
2. It fixes some Windows warnings in the source code. Most importantly, it fixes lots of dll warnings by adjusting C10_API to TORCH_API or TORCH_PYTHON_API. There are still some dll warnings because some TORCH_API functions are actually built as part of libtorch_python
 
cc @malfet @seemethere @peterjc123 @mszhanyi @skyline75489 @nbcsm @EikanWang @jgong5 @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

File: aten/src/ATen/core/op_registration/infer_schema.h [hunk 1]
```diff
@@ -156,6 +156,6 @@ FunctionSchema inferFunctionSchemaSingleReturn(std::string&& name, std::string&&
   return detail::infer_schema::createFunctionSchemaFromTraitsSingleReturn<guts::infer_function_traits_t<FuncType>>(std::move(name), std::move(overload_name));
 }
 
-TORCH_API c10::optional<std::string> findSchemaDifferences(const FunctionSchema& inferred, const FunctionSchema& specified);
+C10_API c10::optional<std::string> findSchemaDifferences(const FunctionSchema& inferred, const FunctionSchema& specified);
```
Comment (malfet) [inline aten/src/ATen/core/op_registration/infer_schema.h:159]: Hmm, how this change is related to warning cleanup?

File: c10/util/Flags.h [hunk 2]
```diff
@@ -208,7 +208,7 @@ C10_DECLARE_REGISTRY(C10FlagsRegistry, C10FlagParser, const std::string&);
   C10_DEFINE_typed_var(std::string, name, default_value, help_str)
 
 // DECLARE_typed_var should be used in header files and in the global namespace.
-#define C10_DECLARE_typed_var(type, name) C10_IMPORT extern type FLAGS_##name
+#define C10_DECLARE_typed_var(type, name) C10_API extern type FLAGS_##name
```
Comment (malfet) [inline c10/util/Flags.h:211]: Again, why?

File: torch/csrc/jit/tensorexpr/mem_dependency_checker.cpp [hunk 3]
```diff
@@ -204,7 +204,7 @@ void AccessInfo::dumpDOT(std::ostream& os) const {
     os << "n" << id_ << " [\n";
     os << "label = \"" << AccessToString(type_) << " (#" << id_ << ")\\n";
     os << "buf : " << *var_ << "\\n";
-    os << "bounds : \[";
+    os << "bounds : [";
```
Comment (malfet) [inline torch/csrc/jit/tensorexpr/mem_dependency_checker.cpp:207]: +1 though it perhaps can be submitted as super trivial one-liner.

File: aten/src/ATen/core/op_registration/infer_schema.h (hunk 1 above)
Comment (cyyever) [reply aten/src/ATen/core/op_registration/infer_schema.h:159]: To solve a dll inconsistent warning.

File: torch/csrc/jit/runtime/static/ops.h [hunk 4]
```diff
@@ -43,7 +43,7 @@ C10_DECLARE_REGISTRY(SROperatorRegistry, SROperatorFunctor);
   };                                                         \
   C10_REGISTER_CLASS(SROperatorRegistry, name, SROperatorFunctor_##id);
 
-C10_DECLARE_REGISTRY(SRNativeOperatorRegistry, SROperatorFunctor);
+TORCH_DECLARE_REGISTRY(SRNativeOperatorRegistry, SROperatorFunctor);
```
Comment (malfet) [inline torch/csrc/jit/runtime/static/ops.h:46]: Why this change is needed?

File: torch/csrc/jit/python/pybind_utils.h [hunk 5]
```diff
@@ -703,7 +703,6 @@ inline void guardAgainstNamedTensor(const T& var) {
 
 // Defined in pybind_utils.cpp to break a circular dependency with
 // python_ivalue.h
```
Comment (malfet) [inline torch/csrc/jit/python/pybind_utils.h:705]: If above comment is no longer relevant, please remove it as well

File: torch/csrc/Module.cpp [hunk 6]
```diff
@@ -1244,11 +1244,7 @@ class WeakTensorRef {
   }
 };
 
-extern "C"
-#ifdef _WIN32
-    __declspec(dllexport)
-#endif
-        TORCH_API PyObject* initModule();
+extern "C" C10_EXPORT PyObject* initModule();
```
Comment (malfet) [inline torch/csrc/Module.cpp:1247]: Shouldn't it be `TORCH_EXPORT`?

File: c10/util/Flags.h (hunk 2 above)
Comment (cyyever) [reply c10/util/Flags.h:211]: If C10_DECLARE_typed_var is used in libc10 headers, we prefer dllexport rather than dllimport. This change eliminates some dll inconsistent warnings.

**Review by malfet — CHANGES_REQUESTED** [review]
In general, change looks OK to me, but please provide a description explaining why changes outside of CMakeLists.txt are necessary. 
In particular, why symbols declared in `torch/` subfolder are marked with `C10_EXPORT` rather than `TORCH_EXPORT`. I.e. windows is the only platform, where export and import definition matter, so `C10_API` is turned into `_declspec(dllexport)` while C10 is compiled, but into `_declspec(dllimport) if used from torch library. It would be great to keep the same paradigm throughout the codebase

File: torch/csrc/jit/tensorexpr/mem_dependency_checker.cpp (hunk 3 above)
Comment (cyyever) [reply torch/csrc/jit/tensorexpr/mem_dependency_checker.cpp:207]: This is triggered by one MSVC warning

File: torch/csrc/Module.cpp (hunk 6 above)
Comment (cyyever) [reply torch/csrc/Module.cpp:1247]: No such a macro defined, only TORCH_API

File: torch/csrc/jit/runtime/static/ops.h (hunk 4 above)
Comment (cyyever) [reply torch/csrc/jit/runtime/static/ops.h:46]: We need to use TORCH_API outside libc10.

File: torch/csrc/jit/python/pybind_utils.h (hunk 5 above)
Comment (cyyever) [reply torch/csrc/jit/python/pybind_utils.h]: Another commit removes the comment

File: torch/csrc/Module.cpp (hunk 6 above)
Comment (cyyever) [reply torch/csrc/Module.cpp:1247]: No macro called TORCH_EXPORT. Only C10_EXPORT is defined to be __declspec(dllexport). It makes little sense to define TORCH_EXPORT to point to C10_EXPORT for this code snippet.

**Comment by cyyever** [comment]
@malfet More detailed description of this PR is added and some suggestions are adopted.

File: torch/csrc/dynamo/guards.cpp [hunk 7]
```diff
@@ -321,8 +321,8 @@ static PyTypeObject TensorGuardsType = {
 static PyObject* check_type_id(PyObject* dummy, PyObject* args) {
   // faster `lambda obj, expected: id(type(obj)) == expected`
   PyObject* obj;
-  unsigned long expected;
-  if (!PyArg_ParseTuple(args, "Ok", &obj, &expected)) {
+  unsigned long long expected;
+  if (!PyArg_ParseTuple(args, "OK", &obj, &expected)) {
```
Comment (malfet) [inline torch/csrc/dynamo/guards.cpp:325]: Nice fix (cc: @jansel ) though dynamo isn't really useful on Windows.4 

But imo it should be part of separate PR, as code as it is written right now is syntactically correct (although semantically wrong, as object id is 64 bit value. Also, why not use `uint64_t` to make it platform agnostic )

**Review by malfet — APPROVED** [review]
LGTM, but please move dynamo changes to its own PR

File: torch/csrc/dynamo/guards.cpp (hunk 7 above)
Comment (cyyever) [reply torch/csrc/dynamo/guards.cpp:325]: CPython document says unsigned  long long but not  uint64_t, although in practice it must be uint64_t on all major OS.
